narwhal/primary/src/state_handler.rs

```diff
@@ -1,35 +1,201 @@
+}
+
+impl RandomnessState {
+    fn try_new(
+        protocol_config: ProtocolConfig,
+        committee: Committee,
+        private_key: RandomnessPrivateKey,
+    ) -> Option<Self> {
+        if !protocol_config.random_beacon() {
+            info!("random beacon: disabled");
+            return None;
+        }
+
+        let info = committee.randomness_dkg_info();
+        let nodes = info
+            .iter()
+            .map(|(id, pk, stake)| nodes::Node::<EncG> {
+                id: id.0,
+                pk: pk.clone(),
+                weight: *stake as u16,
+            })
+            .collect();
+        let nodes = match nodes::Nodes::new(nodes) {
+            Ok(nodes) => nodes,
+            Err(err) => {
+                error!("Error while initializing random beacon Nodes: {err:?}");
+                return None;
+            }
+        };
+        let (nodes, t) = nodes.reduce(
+            VALIDITY_THRESHOLD
+                .try_into()
+                .expect("VALIDITY_THRESHOLD should fit in u16"),
+            protocol_config.random_beacon_reduction_allowed_delta(),
+        );
+        let party = match dkg::Party::<PkG, EncG>::new(
+            private_key,
+            nodes,
+            t.into(),
+            fastcrypto_tbls::random_oracle::RandomOracle::new(
+                format!("dkg {}", committee.epoch()).as_str(),
+            ),
+            &mut rand::thread_rng(),
+        ) {
+            Ok(party) => party,
+            Err(err) => {
+                error!("Error while initializing random beacon Party: {err:?}");
+                return None;
+            }
+        };
+        info!("random beacon: state initialized");
+        Some(Self {
+            party,
+            messages: Vec::new(),
+            processed_messages: Vec::new(),
+            shares: dkg::SharesMap::with_capacity(0),
+            confirmations: Vec::new(),
+            dkg_output: None,
+        })
+    }
+
+    async fn start_dkg(&self, tx_system_messages: &Sender<SystemMessage>) {
+        let msg = self.party.create_message(&mut rand::thread_rng());
+        info!("random beacon: sending DKG Message: {msg:?}");
+        let _ = tx_system_messages
+            .send(SystemMessage::DkgMessage(msg))
+            .await;
+    }
+
+    fn add_message(&mut self, msg: dkg::Message<PkG, EncG>) {
+        if !self.shares.is_empty() {
+            // We've already sent a `Confirmation`, so we can't add any more messages.
+            return;
+        }
+        self.messages.push(msg.clone());
+        match self.party.process_message(msg, &mut rand::thread_rng()) {
+            Ok(processed) => {
+                self.processed_messages.push(processed);
+            }
+            Err(err) => {
+                debug!("error while processing randomness DKG message: {err:?}");
+            }
+        }
+    }
+
+    fn add_confirmation(&mut self, conf: dkg::Confirmation<EncG>) {
+        self.confirmations.push(conf)
```

Comment:
if self.shares.is_empty() { return }

Reply:
I'm not sure we can necessarily count on this, in case local node missed commits and other nodes are already ahead onto the confirmations before we finish processing messages. Is this somehow necessary for security?

Reply:
I'm missing something here - Isn't `handle_sequenced` being called by all honest validators in the same order, with the same arguments? in other words, don't we call that function once per BS commit, no matter if it's a direct commit or an indirect one?
- If not, then I agree with "I'm not sure we can necessarily count on this" but it means that different validators may call `advance` with different states (e.g., different `self.processed_messages`), no? that could be a problem since the DKG code assumes all honest parties use the same messages.
- If it is the case, then `if self.shares.is_empty() { return }` should be safe, no?

In any case, we could call `advance` in the loop in line 262 after each message, to make sure all validators process all messages in the same order (`advance` is the state machine processor, so it might make sense to call it after every input in any case).

Reply:
It's unclear to me (without digging in more to check) whether it's called once per BS commit, or once per direct commit. I didn't realize it was a requirement that all honest parties use the same messages. Shouldn't it be fine as long as they all share a subset of messages that add up to more than the threshold?

It seemed better to me not to call advance every time, because if we have a whole commit's worth of messages available that don't require any additional waiting, it would be good to use all of them, instead of cutting off DKG *right* at the threshold?

Reply:
Resolved per offline discussion to call advance in the loop each cert